src/main.rs

```diff
@@ -1,98 +1,173 @@
-use std::convert::TryInto;
+use std::env;
 
-const BLOCK_SIZE: usize = 32; // Size of each block in bits
-const HASH_SIZE: usize = 32; // Size of the hash code in bits
+const BLOCK_SIZE: usize = 8; // Size of each block in bytes
+const HASH_SIZE: usize = 8; // Size of the hash code in bytes
+const INITIAL_STATE: [u8; HASH_SIZE] = [255, 100, 211, 37, 112, 167, 41, 37]; // randomly generated initial state
 
-struct XorHasher {
+struct EvanHash {
     state: [u8; HASH_SIZE],
-    block_count: usize,
+    block: [u8; BLOCK_SIZE],
+    length: usize,
+    data: Vec<u8>,
+    debug: bool,
 }
 
-impl XorHasher {
-    fn new() -> Self {
-        XorHasher {
-            state: [0; HASH_SIZE],
-            block_count: 0,
-        }
+impl EvanHash {
+    fn new(data: &[u8], debug: bool) -> Self {
+      let hasher = EvanHash {
+        state: INITIAL_STATE,
+        block: [0u8; BLOCK_SIZE],
+        length: 0,
+        data: Vec::from(data),
+        debug: debug,
+      };
+      hasher
     }
 
-    fn update(&mut self, data: &[u8]) {
-        let mut offset = 0;
-
-        while offset < data.len() {
-            let mut block = [0u8; BLOCK_SIZE];
-            let remaining = data.len() - offset;
-            let block_size = remaining.min(BLOCK_SIZE);
+    fn update(&mut self) {
+        self.block = [0u8; BLOCK_SIZE];
+        let remaining = self.data.len() - self.length;
+        let end = self.length + remaining.min(BLOCK_SIZE);
 
-            block[..block_size].copy_from_slice(&data[offset..offset + block_size]);
-            self.process_block(&block);
+        // fill the block from input data
+        self.block[..remaining.min(BLOCK_SIZE)].copy_from_slice(&self.data[self.length..end]);
 
-            offset += block_size;
-            self.block_count += 1;
+        // if this is the last block append data.len()
+        if remaining < BLOCK_SIZE {
+            self.block[remaining] = self.data.len() as u8;
```

Comment:
Length rolls over at 256. It is probably significantly more secure with a larger length. SHA2 uses u64. For EvanHash this would always require an extra block for length.